proposals/build_tarball.md

```diff
@@ -0,0 +1,43 @@
+# Proposal: Build Image Tarball
+
+## Motivation
+
+Currently, Jib can build a container image to either a registry or a Docker daemon. Recently, there
+have been requests for the ability to build an image tarball directly to the filesystem so that the
+user may load it into a Docker daemon via `docker load`, either manually or via a build system.
+
+## Goals
+
+* Build the image and output as a tarball
+* Minimize extra configuration
+
+## Constraints
+
+* The tarball will be saved to the project's build output directory with the name `jib-image.tar`
+
+## Intended Workflow
+
+The user will be able to output a `docker load`able image tarball by running `gradle jibBuildTar`
+for Gradle or `mvn jib:buildTar` for Maven.
+
+## Implementation
+
+The "build tarball" task is almost identical to the "build to docker daemon" task, except for the
+final step, which writes the final tarball to a file instead of piping it to a `docker load`
+command. To avoid duplicate code, a boolean parameter can be used in the steps runners to determine
+whether to build to a Docker daemon or build to a tarball.
+
+The following changes will be made to the code:
+1. Add a private output path configuration parameter to `BuildConfiguration`
+2. Rename:
+   - `BuildTarballAndLoadDockerStep` to `BuildTarballStep`
+   - `runBuildTarballAndLoadDockerStep` to `runBuildTarballStep`
+   - `waitOnBuildTarballAndLoadDockerStep` to `waitOnBuildTarballStep`
+3. Add a boolean parameter `doDockerLoad` to `runBuildTarballStep` and `BuildTarballStep`'s
```

Comment:
build steps probably should not have inputs that change their behavior. For now, to me, it seems like it would make sense to do this as a new static factory on BuildSteps that is just `forBuildToDockerDaemon` without the final step of `docker load`

I guess this is a good argument for choosing a final step and building the steps tree backwards from there? So we can potentially plug in to any random step anywhere as a build target?

Reply:
That second paragraph doesn't really have anything to do with this proposal, it's about the underlying mechanism. So you can ignore it.

Reply:
The way it is right now, the build to docker daemon doesn't actually generate the tar file on disk, it just pipes the tarball data directly to a `docker load` process. The tarball generation + piping to `docker load` happen in the same step at the moment, so it wouldn't really be a matter of just removing the last step, it would mean splitting up the last step into two parts and adding an extra step for generating the file.